src/node_api_embedding.cc

```diff
@@ -0,0 +1,765 @@
+        exec_args_(EmbeddedPlatform::GetInstance()->exec_args()) {}
+
+  EmbeddedEnvironmentOptions(const EmbeddedEnvironmentOptions&) = delete;
+  EmbeddedEnvironmentOptions& operator=(const EmbeddedEnvironmentOptions&) =
+      delete;
+
+  bool is_frozen_{false};
+  node_api_env_flags flags_{node_api_env_default_flags};
+  std::vector<std::string> args_;
+  std::vector<std::string> exec_args_;
+  node::EmbedderPreloadCallback preload_cb_{};
+  node::EmbedderSnapshotData::Pointer snapshot_;
+  std::function<void(const node::EmbedderSnapshotData*)> create_snapshot_;
+  node::SnapshotConfig snapshot_config_{};
+};
+
+struct IsolateLocker {
+  IsolateLocker(node::CommonEnvironmentSetup* env_setup)
+      : v8_locker_(env_setup->isolate()),
+        isolate_scope_(env_setup->isolate()),
+        handle_scope_(env_setup->isolate()),
+        context_scope_(env_setup->context()) {}
+
+  bool IsLocked() const {
+    return v8::Locker::IsLocked(v8::Isolate::GetCurrent());
+  }
+
+  void IncrementLockCount() { ++lock_count_; }
+
+  bool DecrementLockCount() {
+    --lock_count_;
+    return lock_count_ == 0;
+  }
+
+ private:
+  int32_t lock_count_ = 1;
+  v8::Locker v8_locker_;
+  v8::Isolate::Scope isolate_scope_;
+  v8::HandleScope handle_scope_;
+  v8::Context::Scope context_scope_;
+};
+
+class EmbeddedEnvironment final : public node_napi_env__ {
+ public:
+  explicit EmbeddedEnvironment(
+      std::unique_ptr<EmbeddedEnvironmentOptions>&& env_options,
+      std::unique_ptr<node::CommonEnvironmentSetup>&& env_setup,
+      v8::Local<v8::Context> context,
+      const std::string& module_filename,
+      int32_t module_api_version)
+      : node_napi_env__(context, module_filename, module_api_version),
+        env_options_(std::move(env_options)),
+        env_setup_(std::move(env_setup)) {
+    env_options_->is_frozen_ = true;
+
+    std::scoped_lock<std::mutex> lock(shared_mutex_);
+    node_env_to_node_api_env_.emplace(env_setup_->env(), this);
+  }
+
+  static node_napi_env GetOrCreateNodeApiEnv(node::Environment* node_env) {
+    std::scoped_lock<std::mutex> lock(shared_mutex_);
+    auto it = node_env_to_node_api_env_.find(node_env);
+    if (it != node_env_to_node_api_env_.end()) return it->second;
+    // TODO: (vmoroz) propagate API version from the root environment.
+    node_napi_env env = new node_napi_env__(
+        node_env->context(), "<worker_thread>", NAPI_VERSION_EXPERIMENTAL);
+    node_env->AddCleanupHook(
+        [](void* arg) { static_cast<node_napi_env>(arg)->Unref(); }, env);
+    node_env_to_node_api_env_.try_emplace(node_env, env);
+    return env;
+  }
+
+  static EmbeddedEnvironment* FromNapiEnv(napi_env env) {
+    return static_cast<EmbeddedEnvironment*>(env);
+  }
+
+  node::CommonEnvironmentSetup* env_setup() { return env_setup_.get(); }
+
+  std::unique_ptr<node::CommonEnvironmentSetup> ResetEnvSetup() {
+    return std::move(env_setup_);
+  }
+
+  napi_status OpenScope() {
```

Comment:
I assume that a scope is something different than a handle_scope - https://nodejs.org/api/n-api.html#napi_handle_scope, just wondering if there might be confusion between the concepts?

Reply:
Yes, it is different. When we are inside of a module we already have some current `v8::Isolate` and `v8::Context`. We do not have them when we are outside and operating with the environment. So, we must establish them to use any V8/Node API. In the standalone v8-jsi project I used a function [`jsr_run_task`](https://github.com/microsoft/v8-jsi/blob/27e6446af03ef9ebc226c0ef30ef04b96b136228/src/node-api/js_runtime_api.h#L194) that opens/closes the scope internally.